src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/LanguageServices/EditorConfigFileWatcher.cs

```diff
@@ -0,0 +1,271 @@
+        [ConfiguredProjectAutoLoad(RequiresUIThread = false)]
+#pragma warning restore RS0030 // Do not used banned APIs
+        [AppliesTo(ProjectCapability.DotNet)]
+        internal void Load()
+        {
+            var _ = InitializeAsync();
+        }
+
+        /// <summary>
+        /// Called on changes to the project tree.
+        /// </summary>
+        internal async Task DataFlow_ChangedAsync(IProjectVersionedValue<Tuple<IProjectTreeSnapshot, IProjectSubscriptionUpdate>> dataFlowUpdate)
+        {
+            await InitializeAsync();
+
+            IProjectTreeSnapshot treeSnapshot = dataFlowUpdate.Value.Item1;
+            IProjectTree newTree = treeSnapshot.Tree;
+            if (newTree == null)
+            {
+                return;
+            }
+
+            // If tree changed when we are disposing then ignore the change.
+            if (IsDisposing)
+            {
+                return;
+            }
+
+            IProjectSubscriptionUpdate projectUpdate = dataFlowUpdate.Value.Item2;
+            SortedSet<string> updatedEditorConfigFilePaths = GetEditorConfigFilePaths(newTree, projectUpdate);
+
+            var filesToStopWatching = _filesBeingWatched.Keys.Where(p => !updatedEditorConfigFilePaths.Contains(p)).ToArray();
+            await UnregisterFileWatchersAsync(filesToStopWatching);
+
+            var filesToStartWatching = updatedEditorConfigFilePaths.Where(p => !_filesBeingWatched.TryGetValue(p, out var _)).ToArray();
+            await RegisterFileWatchersAsync(filesToStartWatching);
+        }
+
+        private async Task UnregisterFileWatchersAsync(string[] filesToStopWatching)
+        {
+            IVsAsyncFileChangeEx fileChangeService = await _fileChangeService.GetValueAsync();
+            foreach (string editorConfigFilePath in filesToStopWatching)
+            {
+                uint cookie = _filesBeingWatched[editorConfigFilePath];
+                if (cookie != VSConstants.VSCOOKIE_NIL)
+                {
+                    await fileChangeService.UnadviseFileChangeAsync(cookie);
+                }
+
+                _filesBeingWatched.Remove(editorConfigFilePath);
+            }
+        }
+
+        private async Task RegisterFileWatchersAsync(string[] filesToStartWatching)
+        {
+            IVsAsyncFileChangeEx fileChangeService = await _fileChangeService.GetValueAsync();
+            foreach (string editorConfigFilePath in filesToStartWatching)
+            {
+                _filesBeingWatched[editorConfigFilePath] =
+                    await fileChangeService.AdviseFileChangeAsync(
+                        editorConfigFilePath,
+                        _VSFILECHANGEFLAGS.VSFILECHG_Add | _VSFILECHANGEFLAGS.VSFILECHG_Del,
+                        sink: this);
+            }
+        }
+
+        /// <summary>
+        /// Initialize the watcher.
+        /// </summary>
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
```

Comment:
I believe this whole method is overly complicated, but I'm not sure what the simpler version looks like. Notably, I don't actually care about the .editorconfig files passed to the compiler, and within this method I don't even care that files have changed in the general case.

All I really need to be able to do is identify the set of .editorconfig files above the project cone when the project is created/opened and when it is moved. Other changes are not relevant.

Reply:
Projects are never moved, the can be renamed, but their directory always remains the same.

I'm not sure I understand the created/opened scenario, maybe we can talk tomorrow to discuss what you are trying to do? TBH, it's not clear why adding a file in a parent directory is different from adding file under a project.

Reply:
Experimentally, dropping a .editorconfig file into project's directory (or a sub-directory) causes a re-evaluation/design-time build without me having to do anything else. I've been assuming CPS is just watching for any changes under the project directory.

Reply:
Only if its included in a glob.

Reply:
Let's sync up on this so that we can chat about the design.